tfhe/src/integer/server_key/radix/shift.rs

```diff
@@ -117,50 +117,82 @@ impl ServerKey {
     /// sks.unchecked_scalar_right_shift_assign(&mut ct, shift);
     ///
     /// // Decrypt:
-    /// let dec: u64 = cks.decrypt(&ct);
+    /// let dec = cks.decrypt(&ct);
     /// assert_eq!(msg >> shift, dec);
     /// ```
     pub fn unchecked_scalar_right_shift_assign<PBSOrder: PBSOrderMarker>(
         &self,
         ct: &mut RadixCiphertext<PBSOrder>,
-        shift: usize,
+        shift: u64,
     ) {
-        let tmp = self.key.message_modulus.0 as f64;
-
-        //number of bits of message
-        let nb_bits = tmp.log2() as usize;
+        // see parallel implementation for a bit more details
 
-        // 2^u = 2^{p*q+r} = 2^{p*(q+1)}*2^{r-p}
-        let quotient = shift / nb_bits;
+        debug_assert!(ct.block_carries_are_empty());
+        debug_assert!(self.key.carry_modulus.0 >= self.key.message_modulus.0 / 2);
 
-        //p-r
-        let modified_remainder = nb_bits - (shift % nb_bits);
+        if shift == 0 {
+            return;
+        }
 
-        //if r == 0
-        if modified_remainder == nb_bits {
-            self.full_propagate(ct);
-            self.blockshift_right_assign(ct, quotient);
-        } else {
-            // B/2^u = (B*2^{p-r}) / (2^{p*(q+1)})
-            self.unchecked_scalar_left_shift_assign(ct, modified_remainder);
+        let num_bits_in_block = (self.key.message_modulus.0 as f64).log2() as u64;
```

Comment:
can self.key.message_modulus.0.ilog2() be enough ? it's the floor of the log2 though

Reply:
It should work, but it would bump the rust requirement to 1.67

Reply:
ah damn, I think I put some ilog2 somewhere 🤦‍♂️

Reply:
we need to upgrade our rust version -_-

Reply:
So i can put it